### PR TITLE
Feat : Add a scheduling helper for orbax dags.

### PR DESCRIPTION
**Pull request by lechipai0209**
# Description

This PR adds a scheduling helper for Orbax DAGs to centralize schedule calculation and validation.

The helper ensures DAG schedules do not overlap within a 24-hour window and validates `dagrun_timeout` constraints. Unit and CI tests are included to cover both normal and edge cases.

# Example

1. Register your dag in class `SchedulingHelper` registry, indicate relative path of the project and the cluster it would use.
```
  registry: dict[str, Project] = {
      "orbax": Project("dags/orbax/", XpkClusters.TPU_V5P_128_CLUSTER),
  }
```
2. use `ArrangeScheduleTime` in other file. Indicate corresponding registry info and dag_id, it would return an cron expression or error message.
```
  schedule = SchedulingHelper.ArrangeScheduleTime(
      SchedulingHelper.registry["orbax"],
      "maxtext_regular_restore_with_node_disruption",
  )
```

# Tests

- File Path: `dags/common/scheduling_helper/tests/scheduling_helper_test.py`
- Data : `dags/common/scheduling_helper/tests`
- Command : `python -m unittest discover dags/common/scheduling_helper/tests -p "scheduling_helper_test.py"`

| Test case | Input | Output | Description |
|---------|---------|--------|-------------|
| test_arrangescheduletime_is_correct | DAG: "normal_dag" registry, DAG ID: "normal_dag_1" | Schedule string: "0 13 * * *" | Verify that a normal DAG is scheduled correctly according to `ArrangeScheduleTime`. |
| test_nonexist_dag | DAG: "normal_dag" registry, DAG ID: "nonexist_dag" | Raises `ValueError` with message containing "Dag doesn't exist" | Ensure that requesting a non-existent DAG raises the correct exception. |
| test_overtime_dag | DAG: "overtime_dag" registry, DAG ID: "overtime_dag_2" | Raises `ValueError` with message containing "Schedule exceeds 24 hours window;adjust the DEFAULT_MARGIN or dagrun_timeout accordingly." | Check that scheduling a DAG that would exceed 24 hours raises an exception. |
| test_dag_without_timeout | DAG: "no_timeout_dag" registry, DAG ID: "no_timeout_dag" | Raises `ValueError` with message containing "Dag rundag_timeout parameter is necessary" | Ensure that DAGs without a `dagrun_timeout` are rejected. |
| test_all_dags_correctly_scheduled | DAG: "normal_dag" registry, DAG IDs: ["normal_dag_1", "normal_dag_2"] | Schedule strings: ["0 13 * * *", "15 14 * * *"] | Test that all DAGs in the folder are scheduled correctly. |
| test_schedule_replicated | `MockSchedulingHelper.registry` with cluster names | No duplicates (`first_duplicate is None`) | Verify that no duplicate clusters are registered in the registry. |
# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run one-shot tests and provided workload links above if applicable. 
- [X] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Review by alfredyu-cienet — COMMENTED** [review]
please find the existing unit-tests in the repo and implement the similar unit-test(s)

**Review by alfredyu-cienet — COMMENTED** [review]
I think the prototype is ready, next, beyond the tests you implement, let's also do these

1. `dags/common/scheduling_helper_test.py`
2. integrate the unittest with repo's current unittest CI
3. test and validate the schedules of dag registry (not a mock data)
4. 2+3 = every PR trigger the schedule check, using official dag registry
5. test the output of the helper are valid cron format
6. additional mechanism to scan files under a certain folder (say `orbax/`)(reserve capability for future extension) and collect all dags, and make sure all dags are scheduled by helper

**Review by alfredyu-cienet — COMMENTED** [review]
let's take a look at `DagBag` under `airflow.models`, see if that module works better for our scenario

File: xlml/utils/scheduling_helper_test.py
Comment (alfredyu-cienet) [inline xlml/utils/scheduling_helper_test.py]: why duplicate the file?

File: dags/common/scheduling_helper.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    DEFAULT_MARGIN: The default margin time between DAG runs.
+    DEFAULT_ANCHOR: The default anchor time to start scheduling.
+    registry: A mapping from XpkClusterConfig to a list of DAGs associated
+      with the cluster.
+  """
+
+  DEFAULT_MARGIN = dt.timedelta(minutes=15)
+  DEFAULT_ANCHOR = dt.datetime(2000, 1, 1, 13, 0, 0, tzinfo=dt.timezone.utc)
+
+  registry: dict[XpkClusterConfig, list[Dag]] = {
```
Comment (alfredyu-cienet) [inline dags/common/scheduling_helper.py]: for the registry, other than dag registry, we also need the folder registry, since now we'll be checking dags under a certain fold are all registered

File: dags/common/scheduling_helper_test.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+  return dagbags.dag_ids
+
+
+class TestSampleSchedulingHelper(absltest.TestCase):
+  """
+  Test cases for the SchedulingHelper class.
+  """
+
+  def test_schedule_success(self):
```
Comment (alfredyu-cienet) [inline dags/common/scheduling_helper_test.py]: use the same algo to test if the algo run success is meaningless

File: dags/common/scheduling_helper.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  dag_id: str
+  dag_run_timeout: dt.timedelta
+
+  def __init__(
+      self,
+      dag_id: str,
+      dag_run_timeout: dt.timedelta = dt.timedelta(minutes=60),
+  ):
+    self.dag_id = dag_id
+    self.dag_run_timeout = dag_run_timeout
```
Comment (alfredyu-cienet) [inline dags/common/scheduling_helper.py]: I think we can remove this, since `DagBag` collects the dag instances, and we can retrieve the timeout config from `dag.dagrun_timeout`

and furthermore, if we end up with storing the dag_id along, we might as well get rid of this class

Comment (alfredyu-cienet) [inline dags/common/scheduling_helper.py]:
```suggestion
@dataclasses.dataclass
class Dag:
  """
  The metadata of a DAG.
  Attributes:
    dag_id: The DAG ID.
    dag_run_timeout: The maximum allowed runtime of a DAG run.
  """

  dag_id: str
  dag_run_timeout: dt.timedelta = dt.timedelta(minutes=60)
```

File: xlml/utils/scheduling_helper_test.py
Comment (lechipai0209) [reply xlml/utils/scheduling_helper_test.py]: I am a little confused about CI flow. `python3 -m unittest discover xlml "*_test.py"` means it will only scan xlml folder. Also, I found other *_test.py files are put in xlml folder. How can make it right if I only put the test file in common folder ? 

thanks.
